streams/src/main/java/org/apache/kafka/streams/kstream/internals/OuterStreamJoinStoreFactory.java

```diff
@@ -0,0 +1,188 @@
+import org.apache.kafka.streams.state.internals.TimestampedKeyAndJoinSideSerde;
+
+public class OuterStreamJoinStoreFactory<K, V1, V2> extends AbstractConfigurableStoreFactory {
+
+    private final String name;
+    private final StreamJoinedInternal<K, V1, V2> streamJoined;
+    private final JoinWindows windows;
+    private boolean loggingEnabled;
+
+    public enum Type {
+        RIGHT,
+        LEFT
+    }
+
+    public OuterStreamJoinStoreFactory(
+            final String name,
+            final StreamJoinedInternal<K, V1, V2> streamJoined,
+            final JoinWindows windows,
+            final Type type
+    ) {
+        super(streamJoined.dslStoreSuppliers());
+        this.name = buildOuterJoinWindowStoreName(streamJoined, name, type) + "-store";
+        this.streamJoined = streamJoined;
+        this.windows = windows;
+        this.loggingEnabled = streamJoined.loggingEnabled();
+    }
+
+    @Override
+    public StateStore build() {
+        final Duration retentionPeriod = Duration.ofMillis(retentionPeriod());
+        final Duration windowSize = Duration.ofMillis(windows.size());
+        final String rpMsgPrefix = prepareMillisCheckFailMsgPrefix(retentionPeriod, "retentionPeriod");
+        final long retentionMs = validateMillisecondDuration(retentionPeriod, rpMsgPrefix);
+        final String wsMsgPrefix = prepareMillisCheckFailMsgPrefix(windowSize, "windowSize");
+        final long windowSizeMs = validateMillisecondDuration(windowSize, wsMsgPrefix);
+
+        if (retentionMs < 0L) {
+            throw new IllegalArgumentException("retentionPeriod cannot be negative");
+        }
+        if (windowSizeMs < 0L) {
+            throw new IllegalArgumentException("windowSize cannot be negative");
+        }
+        if (windowSizeMs > retentionMs) {
+            throw new IllegalArgumentException("The retention period of the window store "
+                    + name + " must be no smaller than its window size. Got size=["
+                    + windowSizeMs + "], retention=[" + retentionMs + "]");
+        }
+
+        final TimestampedKeyAndJoinSideSerde<K> timestampedKeyAndJoinSideSerde = new TimestampedKeyAndJoinSideSerde<>(streamJoined.keySerde());
+        final LeftOrRightValueSerde<V1, V2> leftOrRightValueSerde = new LeftOrRightValueSerde<>(streamJoined.valueSerde(), streamJoined.otherValueSerde());
+
+        // TODO: we should allow for configuration of this store explicitly instead of assuming that it should
+        // share the same type of store as thisStoreSupplier
+        final boolean useInMemoryStore = streamJoined.thisStoreSupplier() != null
+                && streamJoined.thisStoreSupplier() instanceof InMemoryWindowBytesStoreSupplier;
+        final KeyValueBytesStoreSupplier supplier = useInMemoryStore
+                ? Stores.inMemoryKeyValueStore(name)
+                : dslStoreSuppliers().keyValueStore(new DslKeyValueParams(name));
```

Comment:
I think this is going to resolve to the wrong store type in two cases where it should, at least in theory, be possible to select the right one:

1) Users sets the global `dsl.store.suppliers` (ie the config) to an in-memory or custom store, but want to fix this stream-stream join to rocksdb. So `streamJoined.storeSupplier` == RocksdbBytesStoreSupplier
2) User sets a global `dsl.store.suppliers` and does not configure the StreamJoined with a `dslStoreSupplier`

Frankly 1) is probably not a big deal, but 2) seems like a more serious violation. It's stemming from the fact that we always initialize the `dslStoreSuppliers` in StreamJoined to the default `Materialized.StoreType.ROCKS_DB` if not explicitly configured. So the `streamJoined#dslStoreSuppliers` that we pass into the `AbstractConfigurableStoreFactory` constructor is never null, which means we'll always hit the `Materialized.StoreType.ROCKS_DB` and won't respect the user's global config

At least, that's my reading of the situation...it's a bit complex so maybe I missed something.

Seems like we'll need to either use a flag to mark when the user has explicitly set the StreamJoined's `dslStoreSuppliers`, or just leave it null if the user hasn't set it. The latter feels like the right thing to do but of course that has implications for the rest of the code. Can we make it work without initializing to the default RocksdbDslStoreSuppliers? 

You know the code better than me so I'll leave that up to you, hopefully I got the point across here. At the very least, I guess we could just treat the RocksDB as a special case, and ignore the `StreamJoined#dslStoreSuppliers` in favor of the global default if it's equal to `Materialized.StoreType.ROCKS_DB`

Reply:
> It's stemming from the fact that we always initialize the dslStoreSuppliers in StreamJoined to the default Materialized.StoreType.ROCKS_DB if not explicitly configured.

Ah good catch! I should be leaving that as null as the call to `configure` will properly resolve this in `AbstractConfigurableStoreFactory`.

Reply:
Fixed and added a test for it as well

Reply:
Thanks!

Reply:
One more thing: since we leave the `streamJoined.thisStoreSupplier` as null if not specified (rather than, say, defaulting it to a rocksdb supplier), then we should be able to differentiate when  `thisStoreSupplier` was explicitly set, and therefore should respect when it is set to a RocksDBBytesStoreSupplier. In other words, we can remove the in-memory specific handling and just check if `streamJoined.thisStoreSupplier() != null` and use it if so

That said....I actually think it would be more correct to select the store type according to this ordering:
1. if the `StreamJoined#dslStoreSuppliers` is non-null, use this to get the kv store
2. if the `StreamJoined#thisStoreSupplier` is non-null, and is either rocksdb or in-memory, use a store of the same type
3. Otherwise fall back on the `dsl.store.suppliers` config (whether explicitly configured or the default)

That said, if this is going to be a challenge to implement, I do think it's acceptable to swap the priority of 1 & 2. Either way, we should make sure to respect rocksdb as well as in-memory for `thisStoreSupplier`

Reply:
+1 ! Likes the proposed ordering.